# Titanic EDA – Exploratory Data Analysis & Business Insights

**Project 2: Exploratory Data Analysis**
**Dataset:** Cleaned Titanic Dataset (from Project 1)

This notebook analyzes the cleaned Titanic dataset to discover survival patterns, answer business questions, and generate insights and recommendations for a shipping company's management team.

### Setup: Import Libraries

In [6]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## Part 1: Data Understanding

### • Loading the Dataset

In [8]:
df = pd.read_csv('Cleaned_Titanic.csv')

### • Displaying First 5 Rows

In [9]:
df.head()

,passenger_id,name,sex,age,passenger_class,ticket_number,fare,deck,has_cabin_recorded,siblings_spouses_aboard,parents_children_aboard,port_of_embarkation,survived,family_size,is_alone,age_group,fare_category,survived_flag
0,1,"Braund, Mr. Owen Harris",male,22.0,3,A/5 21171,7.25,Unknown,0,1,0,Southampton,No,2,No,Adult,Low Fare,0
1,2,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,PC 17599,71.28,C,1,1,0,Cherbourg,Yes,2,No,Adult,High Fare,1
2,3,"Heikkinen, Miss. Laina",female,26.0,3,STON/O2. 3101282,7.92,Unknown,0,0,0,Southampton,Yes,1,Yes,Adult,Low Fare,1
3,4,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,113803,53.10,C,1,1,0,Southampton,Yes,2,No,Adult,High Fare,1
4,5,"Allen, Mr. William Henry",male,35.0,3,373450,8.05,Unknown,0,0,0,Southampton,No,1,Yes,Adult,Low Fare,0


### • Displaying Last 5 Rows

In [10]:
df.tail()

,passenger_id,name,sex,age,passenger_class,ticket_number,fare,deck,has_cabin_recorded,siblings_spouses_aboard,parents_children_aboard,port_of_embarkation,survived,family_size,is_alone,age_group,fare_category,survived_flag
886,887,"Montvila, Rev. Juozas",male,27.0,2,211536,13.00,Unknown,0,0,0,Southampton,No,1,Yes,Adult,Low Fare,0
887,888,"Graham, Miss. Margaret Edith",female,19.0,1,112053,30.00,B,1,0,0,Southampton,Yes,1,Yes,Teenager,Medium Fare,1
888,889,"Johnston, Miss. Catherine Helen ""Carrie""",female,21.5,3,W./C. 6607,23.45,Unknown,0,1,2,Southampton,No,4,No,Adult,Medium Fare,0
889,890,"Behr, Mr. Karl Howell",male,26.0,1,111369,30.00,C,1,0,0,Cherbourg,Yes,1,Yes,Adult,Medium Fare,1
890,891,"Dooley, Mr. Patrick",male,32.0,3,370376,7.75,Unknown,0,0,0,Queenstown,No,1,Yes,Adult,Low Fare,0


### • Printing Dataset Shape

In [29]:
print(df.shape)

(891, 18)


In [30]:
##OR
print("Rows and Columns:", df.shape)

Rows and Columns: (891, 18)


In [31]:
##OR
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 891
Columns: 18


### • Displaying Column Names

In [27]:
df.columns


Index(['passenger_id', 'name', 'sex', 'age', 'passenger_class',
       'ticket_number', 'fare', 'deck', 'has_cabin_recorded',
       'siblings_spouses_aboard', 'parents_children_aboard',
       'port_of_embarkation', 'survived', 'family_size', 'is_alone',
       'age_group', 'fare_category', 'survived_flag'],
      dtype='str')

In [28]:
##OR
df.columns.tolist()

['passenger_id',
 'name',
 'sex',
 'age',
 'passenger_class',
 'ticket_number',
 'fare',
 'deck',
 'has_cabin_recorded',
 'siblings_spouses_aboard',
 'parents_children_aboard',
 'port_of_embarkation',
 'survived',
 'family_size',
 'is_alone',
 'age_group',
 'fare_category',
 'survived_flag']

### • Checking Data Types (info())

In [20]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   passenger_id             891 non-null    int64  
 1   name                     891 non-null    str    
 2   sex                      891 non-null    str    
 3   age                      891 non-null    float64
 4   passenger_class          891 non-null    int64  
 5   ticket_number            891 non-null    str    
 6   fare                     891 non-null    float64
 7   deck                     891 non-null    str    
 8   has_cabin_recorded       891 non-null    int64  
 9   siblings_spouses_aboard  891 non-null    int64  
 10  parents_children_aboard  891 non-null    int64  
 11  port_of_embarkation      891 non-null    str    
 12  survived                 891 non-null    str    
 13  family_size              891 non-null    int64  
 14  is_alone                 891 non-null

### • Displaying Summary Statistics (describe())

In [32]:
df.describe() ## for numeric columns only

,passenger_id,age,passenger_class,fare,has_cabin_recorded,siblings_spouses_aboard,parents_children_aboard,family_size,survived_flag
count,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000,891.000000
mean,446.000000,29.112458,2.308642,32.204366,0.228956,0.523008,0.381594,1.904602,0.383838
std,257.353842,13.304352,0.836071,49.693414,0.420397,1.102743,0.806057,1.613459,0.486592
min,1.000000,0.400000,1.000000,0.000000,0.000000,0.000000,0.000000,1.000000,0.000000
25%,223.500000,21.500000,2.000000,7.910000,0.000000,0.000000,0.000000,1.000000,0.000000
50%,446.000000,26.000000,3.000000,14.450000,0.000000,0.000000,0.000000,1.000000,0.000000
75%,668.500000,36.000000,3.000000,31.000000,0.000000,1.000000,0.000000,2.000000,1.000000
max,891.000000,80.000000,3.000000,512.330000,1.000000,8.000000,6.000000,11.000000,1.000000


In [33]:
## OR
df.describe(include='all') ## for numeric+categorical columns

,passenger_id,name,sex,age,passenger_class,ticket_number,fare,deck,has_cabin_recorded,siblings_spouses_aboard,parents_children_aboard,port_of_embarkation,survived,family_size,is_alone,age_group,fare_category,survived_flag
count,891.000000,891,891,891.000000,891.000000,891,891.000000,891,891.000000,891.000000,891.000000,891,891,891.000000,891,891,891,891.000000
unique,NaN,891,2,NaN,NaN,681,NaN,9,NaN,NaN,NaN,3,2,NaN,2,4,3,NaN
top,NaN,"Braund, Mr. Owen Harris",male,NaN,NaN,347082,NaN,Unknown,NaN,NaN,NaN,Southampton,No,NaN,Yes,Adult,Low Fare,NaN
freq,NaN,1,577,NaN,NaN,7,NaN,687,NaN,NaN,NaN,646,549,NaN,537,701,457,NaN
mean,446.000000,NaN,NaN,29.112458,2.308642,NaN,32.204366,NaN,0.228956,0.523008,0.381594,NaN,NaN,1.904602,NaN,NaN,NaN,0.383838
std,257.353842,NaN,NaN,13.304352,0.836071,NaN,49.693414,NaN,0.420397,1.102743,0.806057,NaN,NaN,1.613459,NaN,NaN,NaN,0.486592
min,1.000000,NaN,NaN,0.400000,1.000000,NaN,0.000000,NaN,0.000000,0.000000,0.000000,NaN,NaN,1.000000,NaN,NaN,NaN,0.000000
25%,223.500000,NaN,NaN,21.500000,2.000000,NaN,7.910000,NaN,0.000000,0.000000,0.000000,NaN,NaN,1.000000,NaN,NaN,NaN,0.000000
50%,446.000000,NaN,NaN,26.000000,3.000000,NaN,14.450000,NaN,0.000000,0.000000,0.000000,NaN,NaN,1.000000,NaN,NaN,NaN,0.000000
75%,668.500000,NaN,NaN,36.000000,3.000000,NaN,31.000000,NaN,0.000000,1.000000,0.000000,NaN,NaN,2.000000,NaN,NaN,NaN,1.000000


### • Checking Missing Values

In [24]:
df.isnull().sum()

passenger_id               0
name                       0
sex                        0
age                        0
passenger_class            0
ticket_number              0
fare                       0
deck                       0
has_cabin_recorded         0
siblings_spouses_aboard    0
parents_children_aboard    0
port_of_embarkation        0
survived                   0
family_size                0
is_alone                   0
age_group                  0
fare_category              0
survived_flag              0
dtype: int64

### • Checking Duplicate Values

In [39]:
df.duplicated().sum()

np.int64(0)

In [35]:
#OR
print(f"Number of duplicate rows: {df.duplicated().sum()}")

Number of duplicate rows: 0


### Dataset Description

This dataset contains passenger records from the RMS Titanic, the ship that sank in 1912 after colliding with an
iceberg. The cleaned dataset contains **891 rows and 13 columns**, each row representing one passenger. The
features include a mix of **demographic information** (age, sex), **travel details** (passenger class, fare,
ticket number, cabin deck, port of embarkation), **family relationship counts** (siblings/spouses aboard,
parents/children aboard), and the **target variable** (survived: Yes/No). The dataset has already been cleaned in
Project 1 — there are no missing values and no duplicate rows, which makes it well-suited for direct exploratory
analysis. Overall, this dataset gives us enough information to explore how demographic and socio-economic factors
(class, fare, gender, age, family size) related to a passenger's chances of survival.

## Part 2: Feature Engineering

We will create four new features to enrich our analysis:
1. **Family Size** – total family members traveling together (including the passenger)
2. **Is Alone** – whether the passenger traveled without any family
3. **Age Group** – categorical age bracket
4. **Fare Category** – categorical fare bracket

### 1. Family Size
Formula:  
             FamilySize=SibSp+Parch+1


In [47]:
df['Family_Size'] = df['siblings_spouses_aboard'] + df['parents_children_aboard'] + 1
df[['siblings_spouses_aboard', 'parents_children_aboard','Family_Size']].head()

,siblings_spouses_aboard,parents_children_aboard,Family_Size
0,1,0,2
1,1,0,2
2,0,0,1
3,1,0,2
4,0,0,1


**Why we created it & how it helps:**

Family Size combines the separate family-count columns into one intuitive measure representing the total number of family members traveling together, including the passenger. This makes it easier to analyze whether traveling with a larger or smaller family group influenced survival rates compared with passengers traveling alone.

### 2. Is Alone
If FamilySize = 1, then the passenger is traveling alone.

In [48]:
df['Is_alone'] = np.where(df['Family_Size'] == 1, 'Yes', 'No')
df[['Family_Size', 'Is_alone']].head()

,Family_Size,Is_alone
0,2,No
1,2,No
2,1,Yes
3,2,No
4,1,Yes


**Why we created it & how it helps:**

IsAlone identifies whether a passenger was traveling alone or with family/group using a simple Yes/No flag. This makes the information easier to visualize and compare than raw family-size values, allowing us to analyze whether solo travelers had different survival rates from passengers traveling with family

### 3. Age Group 
Create four categories:
- Child
- Teenager
- Adult
- Senior Citizen

In [52]:
def age_group(age):
    if age < 13:
        return 'Child'
    elif age < 20:
        return 'Teenager'
    elif age < 60:
        return 'Adult'
    else:
        return 'Senior Citizen'

df['age_group'] = df['age'].apply(age_group)

df['age_group'] = pd.Categorical(
    df['age_group'],
    categories=['Child', 'Teenager', 'Adult', 'Senior Citizen'],
    ordered=True
)

df[['age', 'age_group']].head(10)

,age,age_group
0,22.0,Adult
1,38.0,Adult
2,26.0,Adult
3,35.0,Adult
4,35.0,Adult
5,25.0,Adult
6,54.0,Adult
7,2.0,Child
8,27.0,Adult
9,14.0,Teenager


**Why we created it & how it helps:**

AgeGroup divides passengers into life-stage categories—children, teenagers, adults, and senior citizens—instead of analyzing individual age values. This makes aggregate comparisons and survival patterns easier to identify, helping us examine whether children were prioritized during evacuation and how survival differed across different life stages.

## 4. Fare Category
One simple way is to divide fares into three ranges.
- Low Fare
- Medium Fare
- High Fare


In [53]:
def fare_category(fare):
    if fare < 15:
        return 'Low Fare'
    elif fare < 40:
        return 'Medium Fare'
    else:
        return 'High Fare'

df['fare_category'] = df['fare'].apply(fare_category)
df['fare_category'] = pd.Categorical(df['fare_category'],
                                      categories=['Low Fare', 'Medium Fare', 'High Fare'],
                                      ordered=True)
df[['fare', 'fare_category']].head()

,fare,fare_category
0,7.25,Low Fare
1,71.28,High Fare
2,7.92,Low Fare
3,53.10,High Fare
4,8.05,Low Fare


**Why we created it & how it helps:**

FareCategory groups passengers according to the ticket fare they paid instead of using individual fare amounts. Since fare is a continuous and highly skewed variable, grouping it into affordability tiers makes comparisons easier and more intuitive. It can also serve as a proxy for socio-economic status, allowing us to analyze whether passengers who paid higher or lower fares had different survival rates.

### Helper Column: Numeric Survival Flag

In [54]:
# Numeric version of 'survived' for correlation & aggregation purposes
df['survived_flag'] = (df['survived'] == 'Yes').astype(int)
df[['survived', 'survived_flag']].head()

,survived,survived_flag
0,No,0
1,Yes,1
2,Yes,1
3,Yes,1
4,No,0
